Refactor "srcd/sl.jl" file

TorkelE opened the pull request:
Also does the "src/expression_utils.jl" file, as that primarily support the dsl file (the other relevant file is the chemistry one).

Primarily four types of changes:
- Added comment to code which did not have (primarily older code).
- Rewrite some old code (a lot of stuff in the dsl file is stuff I wrote over 6 years ago, which simply can be rewritten better now).
- Do some organisations for where the options are handled in the main function. Been meaning to do this pretty much since I introduced them, and now when we have all *for the foreseeable future) it seemed like a good time to do it. Now it should also be easier to see where to add new options in the future.
- Rewrite so that the `@reaction_network` and `@network_component` macros reuse minimal amount of code (where as previously it the second was basically ctrl+copied versions of the first.

I also add some tests of all the various ways to create models via the DSL, mostly stuff that was not tested before.

TorkelE commented on src/Catalyst.jl:
```diff
@@ -40,6 +40,7 @@ import ModelingToolkit: check_variables,
 
 import Base: (==), hash, size, getindex, setindex, isless, Sort.defalg, length, show
 import MacroTools, Graphs
+using MacroTools: striplines
```
this function is used frequenty, and even more frequently when debugging code. Figured it made sense to import directly.

TorkelE commented on src/chemistry_functionality.jl:
```diff
@@ -86,7 +86,7 @@ function make_compound(expr)
     # we get something like :([:C, :O]), rather than :([C, O]).
     composition = Catalyst.recursive_find_reactants!(expr.args[3], 1,
-        Vector{ReactantStruct}(undef, 0))
+        Vector{ReactantInternal}(undef, 0))
```
`ReactantInternal` and `ReactionInternal` felt like way better name. Especially now with the `Reaction` struct, having a struct called `ReactionStruct` seemed like it could confuse new users. Since it is used internally only (during the creation of models in the DSL), I figured the name could reflect that.

TorkelE commented on src/expression_utils.jl:
```diff
@@ -2,16 +2,19 @@
-            for i in 1:length(ex.args)
-                ex.args[i] = esc_dollars!(ex.args[i])
-            end
+    # If we do not have an expression: recursion has finished and we return the input.
```
Add comments

TorkelE commented on src/expression_utils.jl:
```diff
@@ -22,28 +25,26 @@ end
+    if !isempty(intersect(forbidden_symbols_error, sym))
+        used_forbidden_syms = intersect(forbidden_symbols_error, sym)
+        error("The following symbol(s) are used as species or parameters: $used_forbidden_syms, this is not permitted.")
+    end
```
Just rewritten nicer than what I could do things before (e.g. before I knew about interpolating into strings). No need to have unnecessarily convoluted code.

TorkelE commented on src/dsl.jl:
```diff
@@ -1,64 +1,4 @@
-"""
-
-### Constant Declarations ###
+### Constants Declarations ###
```
Now starts of by declaring relevant constants and the species macro.

TorkelE commented on src/dsl.jl:
```diff
@@ -110,158 +50,207 @@ end
 
-Generates a [`ReactionSystem`](@ref dsl_description) that encodes a chemical reaction
-network.
+Macro for generating chemical reaction network models. Outputs a [`ReactionSystem`](@ref) structure,
```
The old description was pretty much the first documentation of the package, and over 6 years old. Given the important to the package, updating this one seemed long overdue.

TorkelE commented on src/dsl.jl:
```diff
@@ -110,158 +50,207 @@ end
-macro reaction_network(name::Symbol, ex::Expr)
-    :(complete($(make_reaction_system(
-        MacroTools.striplines(ex); name = :($(QuoteNode(name)))))))
+macro reaction_network(name::Symbol, network_expr::Expr)
```
Basically, there are 4 different types of input (nothing, name, reactions, name + reactions). Each case is handled by a set of functions, which are now called from both `@eaction_network` and `@network_component`.

TorkelE commented on src/dsl.jl:
```diff
@@ -110,158 +50,207 @@ end
-# stoichiometry. Recursion makes it able to handle weird cases like 2(X+Y+3(Z+XY)).
+# stoichiometry. Recursion makes it able to handle weird cases like 2(X + Y + 3(Z + XY)). The 
+# reactants are stored in the `reactants` vector. As the expression tree is parsed, the 
+# stoichiometry is updated and new reactants added.
```
Adds comments to the function, explaining what it does.

TorkelE commented on src/dsl.jl:
```diff
@@ -286,112 +277,88 @@ end
 ### DSL Internal Master Function ###
 
 # Function for creating a ReactionSystem structure (used by the @reaction_network macro).
-function make_reaction_system(ex::Expr; name = :(gensym(:ReactionSystem)))
```
name symbol generation is now handled where the user (don't) provide the name as input.

TorkelE commented on src/dsl.jl:
```diff
@@ -286,112 +277,88 @@ end
+    reactions = get_reactions(reaction_lines)
     variables = vcat(variables_declared, vars_extracted)
-
-    # handle independent variables
```
Part of the organisations so that options are handled together (in two sets, some before and some after the reactions are considered, depending on whichever is possible for which option)

TorkelE commented on src/dsl.jl:
```diff
@@ -286,112 +277,88 @@ end
+    # then the output `ReactionSystem`.
     quote
-        $ps
+        $psexpr
```
just renamed so that everything which is an expression for the code that generates something ends with expr (and not just some of it)

TorkelE commented on src/dsl.jl:
```diff
@@ -286,112 +277,88 @@ end
             Catalyst.make_ReactionSystem_internal(
-                $rxexprs, $tiv, setdiff(union($spssym, $varssym, $compssym), $obs_syms),
-                $pssym; name = $name, spatial_ivs = $sivs, observed = $observed_eqs,
+                $rxsexprs, $tiv, setdiff(union($spsvar, $vsvar, $compsvar), $obs_syms),
```
Stuff which is the name of the variable that contains stuff in code ends with var, stuff which is just a list of symbols with syms.

TorkelE commented on src/dsl.jl:
```diff
@@ -517,98 +488,146 @@ function extract_species_and_parameters!(reactions, excluded_syms)
+# Given the extracted species (or variables) and the option dictionary, creates the 
+# `@species ...` (or `@variables ..`) expression which would declare these.
+# If `key = :variables`, does this for variables (and not species).
+function get_usexpr(us_extracted, options, key = :species; ivs = (DEFAULT_IV_SYM,))
```
make the names more general, as it at some point went from handling just species to species and variables (i.e. unknowns)

TorkelE commented on src/dsl.jl:
```diff
@@ -517,98 +488,146 @@ function extract_species_and_parameters!(reactions, excluded_syms)
         symvec = gensym()
-        ex = quote
-            $symvec = $ex
+        expr = quote
```
in some places we used ex and some expr, here I just made it uniform

TorkelE commented on test/dsl/dsl_basic_model_construction.jl:
```diff
@@ -55,6 +55,65 @@ end
+    isequivalent(rs_empty_1, rs_empty)
+    rs_empty_2 == rs_empty
+    rs_empty_3 == rs_empty
+end
```
WHen I remade how the macor inputs are handled at the top level (nothing, name, reactions, name + reactions) I wanted to add some tests that all of this works for all cases and both macros.

TorkelE commented on test/dsl/dsl_advanced_model_construction.jl:
```diff
@@ -212,8 +212,8 @@ end
 # Checks that repeated metadata throws errors.
 let 
-    @test_throws LoadError @eval @reaction k, 0 --> X, [md1=1.0, md1=2.0] 
-    @test_throws LoadError @eval @reaction_network begin
```
This was never supposed to be a `LoadError` in the first place. There was a variable in the `error` statement that was undeclared, so that yielded an `LoadError`, which was not the true error.

TorkelE commented on test/dsl/dsl_advanced_model_construction.jl:
```diff
@@ -131,7 +131,7 @@ let
     # Line number nodes aren't ignored so have to be manually removed
     Base.remove_linenums!(ex)
-    @test eval(Catalyst.make_reaction_system(ex)) isa ReactionSystem
+    @test eval(Catalyst.make_reaction_system(ex, name = QuoteNode(:name))) isa ReactionSystem
```
adapted to the new `make_reaction_system` function (which won't generate names for the user)

TorkelE commented on test/dsl/dsl_advanced_model_construction.jl:
```diff
@@ -131,7 +131,7 @@ let
     # Line number nodes aren't ignored so have to be manually removed
     Base.remove_linenums!(ex)
-    @test eval(Catalyst.make_reaction_system(ex)) isa ReactionSystem
+    @test eval(Catalyst.make_reaction_system(ex, QuoteNode(:name))) isa ReactionSystem
```
adapted to the new `make_reaction_system` function (which won't generate names for the user)

isaacsas commented on src/dsl.jl:
```diff
@@ -453,33 +422,30 @@ function read_reaction_line(line::Expr)
+        # If the `only_use_rate` metadata was not provided, this must be inferred from the arrow.
         metadata_i = get_tup_arg(metadata, i)
-        if all(arg -> arg.args[1] != :only_use_rate, metadata_i.args)
+        if all(arg.args[1] != :only_use_rate for arg in metadata_i.args)
```
Is there some reason you are switching from the functional style in many places to using a generator? I can understand doing so when using `map` to avoid allocations, but this shouldn't have any impact in that regard.

isaacsas commented on src/dsl.jl:
```diff
@@ -453,33 +422,30 @@ function read_reaction_line(line::Expr)
-            ReactionStruct(get_tup_arg(sub_line, i),
-                get_tup_arg(prod_line, i), get_tup_arg(rate, i), metadata_i))
+        # Extracts substrates, products, and rates for the i'th reaction.
+        subs_i, prods_i, rate_i = get_tup_arg.([subs, prods, rate], i)
```
```suggestion
        subs_i, prods_i, rate_i = get_tup_arg.((subs, prods, rate), i)
```

isaacsas commented on src/dsl.jl:
```diff
@@ -517,98 +486,146 @@ function extract_species_and_parameters!(reactions, excluded_syms)
+function get_rxexpr(rx::ReactionInternal)
+    # Initiates the `Reaction` expression.
+    rate = recursive_expand_functions!(rx.rate)
+    rx_constructor = :(Reaction($rate, [], [], [], []; metadata = $(rx.metadata)))
```
Please revert back to the old initialization. The point of using `nothing` was to avoid allocations. Creating empty arrays of type `Any` like here will result in unneeded allocations I believe (and inside the `Reaction` it will immediately set several of them to `nothing` anyways).

isaacsas commented on src/dsl.jl:
```diff
@@ -733,27 +750,40 @@ function create_differential_expr(options, add_default_diff, used_syms, tiv)
-            isempty(ivs) ||
+            if !isempty(ivs)
                 error("An observable ($obs_name) was given independent variable(s). These should not be given, as they are inferred automatically.")
-            isnothing(defaults) ||
+            end
+            if !isnothing(defaults)
                 error("An observable ($obs_name) was given a default value. This is forbidden.")
-            (obs_name in forbidden_symbols_error) &&
+            end
+            if in(obs_name, forbidden_symbols_error)
                 error("A forbidden symbol ($(obs_eq.args[2])) was used as an observable name.")
```
See comments above about dropping `&&` and `||`.

isaacsas commented on src/dsl.jl:
```diff
@@ -733,27 +750,40 @@ function create_differential_expr(options, add_default_diff, used_syms, tiv)
-            # Error checks.
+            end
             if (obs_name in species_n_vars_declared) && is_escaped_expr(obs_eq.args[2])
+                println("HERE")
```
```suggestion
```

isaacsas commented on src/dsl.jl:
```diff
@@ -768,19 +798,18 @@ function read_observed_options(options, species_n_vars_declared, ivs_sorted)
-                push!(observed_vars.args,
+                ivs_get_expr = :(unique(reduce(vcat,
+                    [arguments(MT.unwrap(dep)) for dep in $dep_var_expr])))
+                sort_func(iv) = findfirst(MT.getname(iv) == ivs for ivs in ivs_sorted)
```
```suggestion
                sort_func(iv) = findfirst(==(MT.getname(iv)),  ivs_sorted)
```

isaacsas commented on src/dsl.jl:
```diff
@@ -794,52 +823,68 @@ function read_observed_options(options, species_n_vars_declared, ivs_sorted)
 
-Generates a single [`Reaction`](@ref) object.
+Generates a single [`Reaction`](@ref) object using a similar syntax as the `@reaction_network`
+macro (but permiting only a single reaction). A more detailed introduction to the syntax can
+be found in the description of `@reaction_network`.
```
More generally, I'd suggest following the approach I mentioned for `@reaction_network` aiming this as a reference on options for users who are looking things up, and referring users to the docs for tutorials / introductions. (Along with keeping some example usage.)

isaacsas reviewed (commented):
For future reference, please try not to reorder functions while also modifying them. It would be a lot easier to look at the diff if you had first rewritten them in this PR, and then in a followup PR reordered. 

(Obviously for a change of like 50 lines this can be ignored, but when there are hundreds of lines of diff showing it makes the review a lot longer when things get moved around too.)

TorkelE commented:
Updated per various changes that have happened since the initial PR. Also incorporated your suggestions from the previous one. Still need to go through it properly, will note when this is actually ready.

TorkelE commented on docs/src/model_creation/dsl_advanced.md:
````diff
@@ -501,6 +500,34 @@ Catalyst.getdescription(rx)
+ex = k*A^2 + t
+rx = @reaction b*$ex*$A, $A --> C
+```
+
````
Previously there was actually no section on this in the DSL. The updated `@reaction` docstring now references this one.

TorkelE commented on src/Catalyst.jl:
```diff
@@ -72,11 +72,6 @@ const CONSERVED_CONSTANT_SYMBOL = :Γ
-    fvars = copy(forbidden_symbols_error)
-    delete!(fvars, :t)
-    fvars
-end
```
Per discussion, this one is not actually used.

TorkelE commented on src/chemistry_functionality.jl:
```diff
@@ -86,7 +86,7 @@ function make_compound(expr)
     # we get something like :([:C, :O]), rather than :([C, O]).
     composition = Catalyst.recursive_find_reactants!(expr.args[3], 1,
-        Vector{ReactantStruct}(undef, 0))
+        Vector{DSLReactant}(undef, 0))
```
Better name so that it is not confused with other structures (thanks for the name suggestion)

TorkelE commented on src/expression_utils.jl:
```diff
@@ -22,33 +25,21 @@ end
-        error("The following symbol(s) are used as variables: " *
-              ((map(s -> "'" * string(s) * "', ",
-                  intersect(forbidden_variables_error, v))...)) *
-              "this is not permitted.")
```
Again, the only time this function is issued it don't actually have any use. Old remnant and thus removed.

TorkelE commented on test/dsl/dsl_advanced_model_construction.jl:
```diff
@@ -123,19 +122,6 @@ let
-    exsys = Catalyst.make_reaction_system(ex)
-    sys = @eval Catalyst $exsys
-    @test sys isa ReactionSystem
-end
```
We can discuss this one, but what it actually does is kind of funny. It basically takes the internal function and evaluates it in the current scope somehow. However, when we have clarified the macro output to make it easier to read, e.g. putting
```julia
observed = $obs_eqs
        continuous_events = $continuous_events_expr
        discrete_events = $discrete_events_expr
```
in the macro and then use
```julia
make_ReactionSystem_internal(rx_eq_vec, $tiv, vars, $psvar; name, spatial_ivs,
                observed, continuous_events, discrete_events, combinatoric_ratelaws)
```
we get the trouble that `observed` overwrites `observed` which is an exported function. Now the test is kind of artificial, so Catalyst works perfectly even if this generates an error, and it makes e.g. `@macroexpand` a bit easier to use. The contrary argument for keeping this is that if one wants to copy and evaluate the code then one must modify. Right now I removed the test, but we can discuss.

TorkelE commented on test/dsl/dsl_options.jl:
```diff
@@ -609,11 +616,11 @@ let
-        #@variables D(t)
-        #@equations D(V) ~ 1 - V
-        #d, X --> 0
-    #end
```
These ones are now fixed. Will close https://github.com/SciML/Catalyst.jl/issues/1173

TorkelE commented on test/dsl/dsl_options.jl:
```diff
@@ -675,7 +682,7 @@ let
     @test plot(sol; idxs=:X).series_list[1].plotattributes[:y][end] ≈ 10.0
     @test plot(sol; idxs=[X, Y]).series_list[2].plotattributes[:y][end] ≈ 3.0
     @test plot(sol; idxs=[rn.X, rn.Y]).series_list[2].plotattributes[:y][end] ≈ 3.0
-    @test plot(sol; idxs=[:X, :Y]).series_list[2].plotattributes[:y][end] ≈ 3.0 # (https://github.com/SciML/ModelingToolkit.jl/issues/2778)
```
Think this link has probably remained since when this was broken long ago.

TorkelE commented on src/dsl.jl:
```diff
@@ -110,159 +50,186 @@ end
+    rs_expr = make_reaction_system(striplines(network_expr), name)
+    complete && (rs_expr = :(complete($rs_expr)))
+    return rs_expr
 end
```
I tried to rewrite the `@reaction_network` and `@network_component` macros so that we don't provide repeat code in the two macros.

TorkelE replied:
`make_rs_expr` now does all the work (with the `complete` argument being `true` or `false`).

TorkelE replied:
There are basically four cases of `@reaction_network` (or `@network_component`) inputs, we have one macro command handling each one separately.

TorkelE commented on src/dsl.jl:
```diff
@@ -110,159 +50,186 @@ end
+        prods = recursive_find_reactants!(prod_line, 1, Vector{DSLReactant}(undef, 0))
         metadata = extract_metadata(metadata_line)
-        new(sub, prod, rate, metadata, rx_line)
+        new(subs, prods, rate, metadata, rx_line)
```
subs and prods because there might be multiple ones,

TorkelE commented on src/dsl.jl:
```diff
@@ -110,159 +50,186 @@ end
+        idx = findfirst(r.reactant == ex for r in reactants)
+        if !isnothing(idx)
+            new_mult = processmult(+, mult, reactants[idx].stoichiometry)
+            reactants[idx] = DSLReactant(ex, new_mult)
```
Instead of one long line that get split into two lines by the line break I just wrote it out into two lines
```julia
new_mult = processmult(+, mult, reactants[idx].stoichiometry)
reactants[idx] = DSLReactant(ex, new_mult)
```

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,113 @@ end
-    # Handle interpolation of variables
+    # Handle interpolation of variables in the input.
     ex = esc_dollars!(ex)
 
```
This main function have now been reordered, trying to put the various option handling into separate sections depending on whether they deal wth declared symbols (must happen first), inferred symbols (happens after that) or other options (mostly must happen at the end). Previously stuff was a little bit all over the place depending on when it got added. Now it should be much clearer what is going on.

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,113 @@ end
+
+    # Assemblies the full expression that declares all required symbolic variables, and
+    # then the output `ReactionSystem`.
+    MacroTools.flatten(striplines(quote
```
Since we already did some additional here to make the output easier to see with `@macroexpand`, I added `MacroTools.flatten(striplines(` as well. I typically always add these manually when doing debugging, and feel they make it way easier to see what is going on.

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,113 @@ end
+        continuous_events = $continuous_events_expr
+        discrete_events = $discrete_events_expr
+        combinatoric_ratelaws = $combinatoric_ratelaws
+        default_reaction_metadata = $default_reaction_metadata
```
I added the remaining options here so that they now all can be looked at. Also renamed them to make the final statement less of a mess (although we might want to reert e.g. `observed`. Trivial ones like `combinatoric_ratelaws = $combinatoric_ratelaws` is technically not needed here.

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,113 @@ end
                 error("Error: Must provide a tuple of reaction rates when declaring a bi-directional reaction.")
-            end
+            (typeof(metadata) != Expr || metadata.head != :tuple) &&
+                error("Error: Must provide a tuple of reaction metadata when declaring a bi-directional reaction.")
```
This should have been here already, but seems it wasn't.

TorkelE commented on src/dsl.jl:
```diff
@@ -516,155 +450,200 @@ end
+    tiv = ivs[1]
+    sivs = (length(ivs) > 1) ? Expr(:vect, ivs[2:end]...) : nothing
+    return tiv, sivs, ivs, ivsexpr
+end
```
Moved this bit out of the main function into a separate function.

TorkelE commented on src/dsl.jl:
```diff
@@ -516,155 +450,208 @@ end
+    eqexpr.args[2] = recursive_escape_functions!(eqexpr.args[2], diffsyms)
+    eqexpr.args[3] = recursive_escape_functions!(eqexpr.args[3], diffsyms)
+    eqexpr
 end
```
Moved `escape_equation!` up here since it is just a helper function to the function above and not involved in other stuff. Calls `recursive_escape_functions!` on the lhs as well.

TorkelE commented on test/dsl/dsl_options.jl:
```diff
@@ -1167,16 +1181,16 @@ end
+    # Test user-defined function (on both lhs and rhs).
     rn = @reaction_network begin
-        @equations D(A) ~ f(A, t)
+        @equations D(A) + g(A) ~ f(A, t)
```
updated because we now hanldes functions on both sides.

TorkelE commented on src/dsl.jl:
```diff
@@ -717,31 +703,38 @@ function read_equations_options(options, syms_declared, parameters_extracted; re
+    if add_default_diff && !any(diff_dec.args[1] == :D for diff_dec in diffsexpr.args)
+        diffs_inferred = [:D]
+        push!(diffsexpr.args, :(D = Differential($(tiv))))
     end
```
Now, if we use a default differential, `diffsexpr` is updated directly in `read_equations_options!`. Made more sense as I wanted to handle the `@differnetials` option at the beginning (since this explicitly declares symbols). Previously this option was handled after `@equations`.

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,113 @@ end
+    spsexpr, spsvar = scalarize_macro(spsexpr_init, "specs")
+    vsexpr, vsvar = scalarize_macro(vsexpr_init, "vars")
+    cmpsexpr, cmpsvar = scalarize_macro(cmpexpr_init, "comps")
+    rxsexprs = get_rxexprs(reactions, equations, union(diffs_declared, diffs_inferred))
```
trying to harmonise notation a bit across everything. exprs is the expression that generates the actually code, var is the generated variable we store stuff in. init on things that should not be evaluated. sps are species, vs variables, us unknowns, ps parameters, cmps compounds. Syms are lists of symbols (e.g. the symbols corresponding to observables). Not insisting we use this across the board, just explaining how I have changed things.

TorkelE commented on src/dsl.jl:
```diff
@@ -516,155 +450,212 @@ end
-    if rateex isa Symbol
-        if !(rateex in forbidden_symbols_skip) && !(rateex in excluded_syms)
-            push!(push_symbols, rateex)
+function add_syms_from_expr!(push_symbols::AbstractSet, expr::ExprValues, excluded_syms)
```
rateex -> expr as this is not necessarily only the rate anymore.

TorkelE commented on src/dsl.jl:
```diff
@@ -516,155 +450,212 @@ end
+        cmps_declared = Union{Symbol, Expr}[]
     end
-    return compound_expr, compound_species
+    return cmpexpr_init, cmps_declared
```
shortened compound to cmp in line with other shortenings, and to avoid having to break certain lines into two lines in some places.

TorkelE commented on src/expression_utils.jl:
```diff
@@ -19,36 +22,9 @@ function is_escaped_expr(expr)
-                  intersect(forbidden_symbols_error, v))...)) *
-              "this is not permitted.")
-    nothing
-end
```
This is not needed, because the check is already carried out in `Reaction` and `ReactionSystem`.

TorkelE commented on src/dsl.jl:
```diff
@@ -914,68 +930,58 @@ end
+    species, parameters = extract_sps_and_ps([reaction], [])
 
-    # Checks for input errors.
-    forbidden_symbol_check(union(species, parameters))
```
This check is already carried out in `Reaction()`, so no longer needed in the DSL.

TorkelE commented on src/expression_utils.jl:
```diff
@@ -22,33 +25,26 @@ end
-    allunique(syms) ||
-        error("Reaction network independent variables, parameters, species, and variables must all have distinct names, but a duplicate has been detected. ")
-    nothing
-end
```
Now only used in one place, there implemented directly.

TorkelE commented on src/expression_utils.jl:
```diff
@@ -22,33 +25,26 @@ end
+    (length(expr.args) > 3) &&
+        error("An option input ($expr) is misformatted. Potentially, it has multiple inputs on a single lines, and these should be split across multiple lines using a `begin ... end` block.")
+    return expr.args[3]
+end
```
Helper function for handling error checks across multiple options.

TorkelE commented on src/spatial_reaction_systems/spatial_reactions.jl:
```diff
@@ -57,9 +57,10 @@ function make_transport_reaction(rateex, species)
-    sexprs = get_sexpr([species], Dict{Symbol, Expr}())
-    pexprs = get_pexpr(parameters, Dict{Symbol, Expr}())
+    sexprs = get_usexpr([species], Dict{Symbol, Expr}())
+    pexprs = get_psexpr(parameters, Dict{Symbol, Expr}())
```
Uses new names in the DSL file

TorkelE commented on test/dsl/dsl_options.jl:
```diff
@@ -415,36 +415,35 @@ let
+# Tests error when disallowed name is used for variable.
+let
+    @test_throws Exception @eval @reaction_network begin
+        @variables π(t)
```
new test

TorkelE commented on src/dsl.jl:
```diff
@@ -1,64 +1,4 @@
-    probODE = ODEProblem(rn, args...; kwargs...)        # Using multiple dispatch the reaction network can be used as input to create ODE, SDE and Jump problems.
-    probSDE = SDEProblem(rn, args...; kwargs...)
-    probJump = JumpProblem(prob,aggregator::Direct,rn)
-"""
```
The first documentation of `@reactio_network`, but now not even a docstring and just a weird remnant. remove.

TorkelE commented on src/dsl.jl:
```diff
@@ -1,64 +1,4 @@
-    probJump = JumpProblem(prob,aggregator::Direct,rn)
-"""
-
-### Constant Declarations ###
```
Not removed, now just line 1 and git got confused.

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,114 @@ end
+    (sum(length, [reaction_lines, option_lines]) != length(ex.args)) &&
+        error("@reaction_network input contain $(length(ex.args) - sum(length.([reaction_lines,option_lines]))) malformed lines.")
+    any(!in(option_keys), keys(options)) &&
+        error("The following unsupported options were used: $(filter(opt_in->!in(opt_in,option_keys), keys(options)))")
```
Step one, read the macro input and check that it is correctly formatted.

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,114 @@ end
+    if !allunique(syms_declared)
+        nonunique_syms = [s for s in syms_declared if count(x -> x == s, syms_declared) > 1]
+        error("The following symbols $(unique(nonunique_syms)) have explicitly been declared as multiple types of components (e.g. occur in at least two of the `@species`, `@parameters`, `@variables`, `@ivs`, `@compounds`, `@differentials`). This is not allowed.")
+    end
```
Step 2, read all options that explicitly declares symbols as stuff.

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,114 @@ end
+        union(syms_declared, sps_inferred), tiv; requiredec)
+    ps_inferred = setdiff(ps_pre_inferred, vs_inferred, diffs_inferred)
+    syms_inferred = union(sps_inferred, ps_inferred, vs_inferred, diffs_inferred)
+    all_syms = union(syms_declared, syms_inferred)
```
Step 3, read the reactions, as well as the options from which we infer symbols.

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,114 @@ end
+    continuous_events_expr = read_events_option(options, :continuous_events)
+    discrete_events_expr = read_events_option(options, :discrete_events)
+    default_reaction_metadata = read_default_noise_scaling_option(options)
+    combinatoric_ratelaws = read_combinatoric_ratelaws_option(options)
```
Step 4, read the other options.

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,114 @@ end
+    spsexpr, spsvar = scalarize_macro(spsexpr_init, "specs")
+    vsexpr, vsvar = scalarize_macro(vsexpr_init, "vars")
+    cmpsexpr, cmpsvar = scalarize_macro(cmpexpr_init, "comps")
+    rxsexprs = get_rxexprs(reactions, equations, all_syms)
```
Step 5, creates the various expressions to put in the macro.

TorkelE commented on src/dsl.jl:
```diff
@@ -516,155 +451,216 @@ end
+        (!isempty(species) && requiredec) &&
+            throw(UndeclaredSymbolicError("Unrecognized reactant $(join(species, ", ")) detected in reaction expression: \"$(string(reaction.rxexpr))\". Since the flag @require_declaration is declared, all species must be explicitly declared with the @species option."))
     end
+    union!(excluded_syms, species)
```
`union!` replaces `foreach(s -> push!(excluded_syms, s), species)`

TorkelE commented on src/dsl.jl:
```diff
@@ -516,155 +451,216 @@ end
-            error("@default_noise_scaling should only have a single input, this appears not to be the case: \"$(options[:default_noise_scaling])\"")
-        end
-        push!(default_reaction_metadata.args,
-            :(:noise_scaling => $(options[:default_noise_scaling].args[3])))
```
`check_default_noise_scaling!` moved later to reflect the order with which these are called.

TorkelE commented on test/dsl/dsl_options.jl:
```diff
@@ -1128,55 +1145,184 @@ let
+    @test_throws Exception @eval @reaction_network begin
+        @combinatoric_ratelaws false false
+        d, 3X --> 0
+    end
```
Checks that new error checks work (don't think the one above actually errored either, but it should)

TorkelE commented on docs/src/model_creation/dsl_advanced.md:
```diff
@@ -575,6 +574,45 @@ nothing # hide
+Here, `@reaction` is followed by a single line consisting of three parts:
+- A rate (at which the reaction occurs).
+- Any number of substrates (which are consumed by the reaction).
+- Any number of products (which are produced by the reaction).
```
This bit
```
Here, `@reaction` is followed by a single line consisting of three parts:
- A rate (at which the reaction occurs).
- Any number of substrates (which are consumed by the reaction).
- Any number of products (which are produced by the reaction).
```

is essentially a repeat of previous documentation. Probably unnecessary for most people and only relevant if we wish to keep this section self contained. happy to remove though.

TorkelE commented on test/dsl/dsl_advanced_model_construction.jl:
```diff
@@ -269,6 +255,24 @@ let
+    @test_throws Exception @eval @reaction_network begin
+        d, X --> 0, [misc=>:misc; description="description"]
+    end
+end
```
One of several new sets of tests for stuff which we previously didn't actually test

TorkelE commented on test/dsl/dsl_basic_model_construction.jl:
```diff
@@ -432,11 +491,74 @@ let
-    @test_throws LoadError @eval @reaction im, 0 --> B
-    @test_throws LoadError @eval @reaction nothing, 0 --> B
-    @test_throws LoadError @eval @reaction k, 0 --> im
-    @test_throws LoadError @eval @reaction k, 0 --> nothing
```
forbidden symbol tests now tested more extensively in a test block two blocks down ("# Check that forbidden symbols correctly generates errors"

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,114 @@ end
+    syms_inferred = union(sps_inferred, ps_inferred, vs_inferred, diffs_inferred)
+    all_syms = union(syms_declared, syms_inferred)
+    obsexpr, obs_eqs, obs_syms = read_observed_option(options, ivs,
+        union(sps_declared, vs_declared), all_syms; requiredec)
```
Step 3: Read all reactions and equations. Finds all symbols that are inferred as something.

TorkelE commented on src/dsl.jl:
```diff
@@ -516,155 +450,197 @@ end
+    eqexpr.args[2] = recursive_escape_functions!(eqexpr.args[2], all_syms)
+    eqexpr.args[3] = recursive_escape_functions!(eqexpr.args[3], all_syms)
+    eqexpr
 end
```
escape equations is called by get_rxexprs

TorkelE commented on src/dsl.jl:
```diff
@@ -914,68 +931,61 @@ end
 
-    # Checks for input errors.
+    # Checks for input errors. Needed here but not in `@reaction_network` as `ReactionSystem` perform this check but `Reaction` don't.
     forbidden_symbol_check(union(species, parameters))
```
For `@reaction_network` this check is not needed as `ReactionSystem` does a similar one. We might consider whether `Reaction` should also perform this type of check (and we remove it from here entirely.

E.g. this works:
```julia
@parameters Γ
@species X(t)
Reaction(Γ, [X], [])
```
while this errors:
```julia
@reaction Γ, X --> 0
```

TorkelE commented:
This should be ready now. However, Plots fails on precompilation, so have to wait until that is fixed.

Not high priority though, doesn't actually change anything (but should make further updates to the DSL easier).

Once this PR is merged, I will have a follow-up one which reorders the option handling functions to follow the order with which they are called.

isaacsas commented on src/dsl.jl:
```diff
@@ -828,83 +805,123 @@ function read_observed_options(options, species_n_vars_declared, ivs_sorted; req
+    if haskey(options, :ivs)
+        ivs = Tuple(extract_syms(options, :ivs))
+        ivsexpr = copy(options[:ivs])
+        ivsexpr.args[1] = Symbol("@", "parameters")
```
```suggestion
        ivsexpr.args[1] = Symbol("@", "parameters")
```
We should probably switch to `@independent_variables` as MTK is going to use that in the future (for now it just wraps `@parameters` but that will change).

TorkelE replied:
Agreed, changing to `@independent_variables`

TorkelE replied:
agreed, I've added this in now

isaacsas reviewed (dismissed):
Please check that non-scalarization is still working ok for vector parameters, and ensure there isn't a performance regression. See the examples in https://github.com/SciML/Catalyst.jl/issues/1051

TorkelE commented on src/dsl.jl:
```diff
@@ -516,155 +450,197 @@ end
-# parameters), and assign them to the right category.
-function extract_species_and_parameters!(reactions, excluded_syms; requiredec = false)
+# parameters) and assign them to the right category.
+function extract_sps_and_ps(reactions, excluded_syms; requiredec = false)
```
This is actually not obvious. Technically, it shouldn't really, it calls `add_syms_from_expr!` which does this. 

However, you are right that it actually does a `union!(excluded_syms, species)` which updates the `excluded_syms` input (which it doesn't hurt to modify, but which is not really an intended effect of  `extract_sps_and_ps`.

We could change
```julia
union!(excluded_syms, species)
```
inside `` to 
```julia
excluded_syms = union(excluded_syms, species)
```
which should prevent `extract_sps_and_ps` from being mutating. But maybe that would have a (very slight) performance impact? Otherwise it would probably be better to reorder its input, `add the `!` and rename `syms_declared` here:
```julia
sps_inferred, ps_pre_inferred = extract_sps_and_ps(reactions, syms_declared; requiredec)
```
(since it is only a vector of declared symbols before it it put into that function)

TorkelE replied:
I have quite a good set of benchmarks lined up for checking DSL (and `ReactionSystem` generally) performance. If the PR shows up fine, I'd be slightly partial to changing to `excluded_syms = union(excluded_syms, species)` (since that is the simplest change), and then do a general look at how we can speed up the core Catalyst parts with some profiling etc.

isaacsas replied:
If you don't want to add a `!` just document that it modifies `excluded_syms` by possibly enlarging it. That should be fine.

TorkelE commented on src/dsl.jl:
```diff
@@ -481,22 +414,23 @@ function read_reaction_line(line::Expr)
+    # The rates, substrates, products, and metadata may be in a tuple form (e.g. `k, (X,Y) --> 0`).
+    # This finds these tuples' lengths (or 1 for non-tuple forms). Inconsistent lengths yield error.
+    lengs = (tup_leng(subs), tup_leng(prods), tup_leng(rate), tup_leng(metadata))
+    maxl = maximum(lengs)
```
Agree that this is better

TorkelE commented on src/dsl.jl:
```diff
@@ -516,155 +450,197 @@ end
+            add_syms_from_expr!(push_symbols, expr.args[i], excluded_syms)
         end
     end
-    nothing
```
Great point, makes lots of sense! Didn't realise, you are right, `nothing` should definitely be here.

TorkelE commented on src/dsl.jl:
```diff
@@ -110,159 +50,187 @@ end
-# Ideally, it would have been possible to combine the @reaction_network and @network_component macros.
-# However, this issue: https://github.com/JuliaLang/julia/issues/37691 causes problem with interpolations
-# if we make the @reaction_network macro call the @network_component macro.
+# Ideally, it would have been possible to combine the @reaction_network and @network_component
+# macros. However, this issue: https://github.com/JuliaLang/julia/issues/37691 causes problem
+# with interpolations if we make the @reaction_network macro call the @network_component macro.
+# Instead, these use the same input, but pass `complete = false` to `make_rs_expr`.
```
Sounds good

TorkelE commented on src/dsl.jl:
```diff
@@ -110,159 +50,187 @@ end
+function make_rs_expr(name; complete = true)
+    rs_expr = :(ReactionSystem(Reaction[], t, [], []; name = $name))
+    complete && (rs_expr = :(complete($rs_expr)))
+    return Expr(:block, :(@parameters t), rs_expr)
```
Agreed

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,114 @@ end
+    options = Dict(Symbol(String(arg.args[1])[2:end]) => arg for arg in option_lines)
+    allunique(arg.args[1] for arg in option_lines) ||
+        error("Some options where given multiple times.")
+    (sum(length, [reaction_lines, option_lines]) != length(ex.args)) &&
+        error("@reaction_network input contain $(length(ex.args) - sum(length.([reaction_lines,option_lines]))) malformed lines.")
```
Since the extra computation was in the error I figured it didn't really matter, and this would be better as it compressed the code one line. But if you prefer this version and `(sum(length, [reaction_lines, option_lines]) != length(ex.args))` cannot be done in a non-allocating manner I am happy to change.

isaacsas replied:
It is just a lot more readable (to me) this way. I can immediately understand the test while the other approach required me to pause and think.

TorkelE replied:
Sounds good

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,114 @@ end
+        # Inserts the expressions which generate the `ReactionSystem` input.
+        $ivsexpr
+        $psexpr
+        $spsexpr
+        $vsexpr
```
Honestly forgotten why I flipped the order, but I agree that species with dpendeant/metadata variables seems more likely than the other way around, will flip it back.

For the serialization code I wrote a functionality which detects symbolic variables that depend on other symbolic variables and declares them in order, at some point we should add that to the DSL as well.
(especially a single pass, where we split symbolic variables with/without dependants would be very easy to implement and clear most cases)

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,114 @@ end
+    # Extract the lines with reactions, the lines with options, and the options. Check for input errors.
     reaction_lines = Expr[x for x in ex.args if x.head == :tuple]
     option_lines = Expr[x for x in ex.args if x.head == :macrocall]
+    options = Dict(Symbol(String(arg.args[1])[2:end]) => arg for arg in option_lines)
+    allunique(arg.args[1] for arg in option_lines) ||
+        error("Some options where given multiple times.")
```
It doesn't really matter. I checked that it works fine. The advantage is that we can throw the error very slightly faster, but never felt that minuscule performance increases really mattered for error throwing, and then it was better to have the code slightly easier to read. We can still flip them though if you think it is important.

TorkelE replied:
sounds good

isaacsas commented on src/dsl.jl:
```diff
@@ -516,155 +450,197 @@ end
+    # If the input expression is non-empty, wrap it with additional information.
+    if expr_init != :(())
+        symvec = gensym()
+        expr = quote
+            $symvec = $expr_init
+            $namesym = reduce(vcat, Symbolics.scalarize($symvec))
```
Please double check the other changes from that PR to the DSL remain in effect (since they are performance related dropping them wouldn't get caught by any test I think).

TorkelE commented on src/dsl.jl:
```diff
@@ -298,149 +266,114 @@ end
+        rx_eq_vec = $rxsexprs
+        us = setdiff(union($spsvar, $vsvar, $cmpsvar), $obs_syms)
+        observed = $obs_eqs
+        continuous_events = $continuous_events_expr
+        discrete_events = $discrete_events_expr
```
That makes sense as the reason. 

Just running stuff like
```julia
using Catalyst, ModelingToolkit
@reaction_network ...
```
is fine. Stuff like this as well:
```julia
using Catalyst, ModelingToolkit
import ModelingToolkit: discrete_events
discrete_events # discrete_events (generic function with 1 method)
rn = @reaction_network begin
    (p,d), 0 <--> X
end
rn = @reaction_network begin
    @discrete_events 1.0 => [X ~ X/2]
    (p,d), 0 <--> X
end
discrete_events # discrete_events (generic function with 1 method)
discrete_events(rn) # OK
```

 The one case I found is if one try to do something like this:
```julia
ex = quote
    (Ka, Depot --> Central)
    (CL / Vc, Central --> 0)
end
Base.remove_linenums!(ex)
name = QuoteNode(:rs)
exsys = Catalyst.make_reaction_system(ex, name)
sys = @eval Catalyst $exsys
```
essentially calling the internal `make_reaction_system` function and trying to execute the code it creates directly. But that feels really niche and weird.

So yeah, I haven't actually been able to find a way where declaring stuff like this causes troubles. However I am no expert in scopes, so maybe there is something. Happy to revert back to the old approach to be on the safe side if you prefer.

isaacsas replied:
If the old approach keeps that example working, which I think maybe people at Pumas were using, let's keep the alternative names (or just add an `_` before the new names to make clear they are internal).

isaacsas replied:
Or we need to figure out how to scope it so we can use the normal names.

TorkelE replied:
If Pumas is using something like
```julia
ex = quote
    (Ka, Depot --> Central)
    (CL / Vc, Central --> 0)
end
Base.remove_linenums!(ex)
name = QuoteNode(:rs)
exsys = Catalyst.make_reaction_system(ex, name)
sys = @eval Catalyst $exsys
```
then we will have to keep it. I thought that `@eval` was a big no no in pretty much any code though?

TorkelE replied:
I have changed to using e.g. `_discrete_events`. This also allows us to reintorudce that test I removed without putting it in a module or something.

isaacsas commented on src/dsl.jl:
```diff
@@ -863,23 +906,23 @@ end
-    # Returns the rephrased expression.
+    # Returns a repharsed expression which generates the `Reaction`.
     quote
         $pexprs
         $iv
```
This never seems to have been addressed?

isaacsas commented:
Ahh, nevermind that comment. For some reason Github showed me an old version of the code... Looks good to me.
